rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp

```diff
@@ -573,6 +585,11 @@ void Rir2Pir::translate(rir::Code* srcCode, Builder& insert,
 
     std::unordered_map<Value*, CallFeedback> callFeedback;
 
+    /* Captures the last type feedback observed. The assumption is that this
```

Comment:
I don't think that this assumption works in general. In particular typefeedback instructions can be inserted anywhere, not all binops might even have one, you might have the typefeedback hoisted. Basically they should be completely independent of the usage. I would prefer if we can collect them in a map like in the call feedback case.

Reply:
So, for this issue I think we have a problem with the current design then. The feedback is stored inside the record_binop instruction. But there is no way to connect a binop with its record_binop. Then when translating the record_binop to pir I may be able to store the feedback in a map on the opcode* or something like that. But then, when translating the binop, there will be no way to find the position in the map. So, according to this design I though there was always a record_binop just before each binop...

Reply:
Currently we have it for this list: https://github.com/reactorlabs/rir/blob/master/rir/src/ir/Compiler.cpp#L197

But I would rather not have the assumption, that a binop must have feedback. That's why I created two separate and independent instructions.

> But there is no way to connect a binop with its record_binop.

of course there is... The record binop records the two values which are top of stack. You map pir `Value*` to type feedback. Baiscally:

```
std::unordered_map<Value*, TypeFeedback&> typeFeedback;
...
case record_binop:
    typeFeedback[at(0)] = typeFeedback[0];
    typeFeedback[at(1)] = typeFeedback[1];
```

then at any point in the compiler,  if you want to know something, you just:

```
if (typefeedback(aValue).noObject())
  doo something
if (typefeeedback(aValue).scalar())
  do something else
```

Reply:
The record_binop is not popping anything from the stack. That is why it seems quite coupled with a following binop concrete operation.  It assume it would not be striaghtforward to hoist this operation much as it is designed right now. 

Anyway, I submitted the change. I agree it is much more robust and will probably keep working if we change how the feedback ops are managed.